transform/plugin.go

```diff
@@ -5,13 +5,40 @@ import (
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 )
 
-type Plugin interface {
+type PluginRun interface {
 	// Determine for a given resources what the plugin is deciding to do with this
-	Run(*unstructured.Unstructured) (PluginResponse, error)
+	Run(*unstructured.Unstructured, map[string]string) (PluginResponse, error)
+}
+
+type Metadata interface {
+	Metadata() (PluginMetadata, error)
+}
+
+type Plugin interface {
+	PluginRun
+	Metadata
 }
 
 type PluginResponse struct {
 	Version    string          `json:"version,omitempty"`
 	IsWhiteOut bool            `json:"isWhiteOut,omitempty"`
 	Patches    jsonpatch.Patch `json:"patches,omitempty"`
 }
+
+type PluginMetadata struct {
+	Name            string
+	Version         string
+	RequestVersion  []Version
+	ResponseVersion []Version
+	OptionalFields  []string
+}
+
+type Version string
+
+const (
+	V1 Version = "v1"
+)
+
+const (
+	MetadataStdIn string = "METADATA"
```

Comment:
Should this be part of the cli package? My interpretation is that the cli implementation of the plugin interface needs to have metadata through stdin right?

Reply:
I think we want to use this string from both the CLI package and the binary-plugin package. Because of this, I think it should be in the plugins.go. WDYT?

Reply:
Sounds reasonable.